dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ORCIDHelper.java

```diff
@@ -150,4 +157,43 @@ private static String transformWork(Work work) throws JAXBException {
         marshaller.marshal(work, writer);
         return writer.getBuffer().toString();
     }
+
+    public static Optional<Long> searchForPutCodeByDoiUrl(String baseURL, String id, List<Token> orcidTokens, String doiUrl)
+            throws IOException, URISyntaxException, InterruptedException, JAXBException {
+        // Get user's ORCID works
+        HttpResponse<String> response = getAllWorks(baseURL, id, orcidTokens.get(0).getToken());
+
+        switch (response.statusCode()) {
+        case HttpStatus.SC_OK:
+            Works works = transformXmlToWorks(response.body());
+            // Find the ORCID work with the DOI URL and get its put code
+            for (WorkGroup work : works.getWorkGroup()) {
+                // There should only be one external ID that matches the DOI url
+                Optional<ExternalID> doiExternalId = work.getIdentifiers().getExternalIdentifier().stream().filter(externalID -> externalID.getValue().equals(doiUrl)).findFirst();
+                if (doiExternalId.isPresent()) {
+                    WorkSummary workSummary = work.getWorkSummary().get(0);
+                    return Optional.of(workSummary.getPutCode());
+                }
+            }
+            return Optional.empty();
+        default:
+            throw new CustomWebApplicationException("Could not export to ORCID: " + response.body(), response.statusCode());
```

Comment:
Maybe add something about not being able to contact the ORCID API (properly) to distinguish this from other failures that can come later